src/storage_versioned/mod.rs

```diff
@@ -0,0 +1,739 @@
+    // Returns Result with StorageVersioned instance or Err with a describing message if some error occurred
+    // NOTE: for storage to be created under 'create_if_missing = true' the specified (in path) storage's directory shouldn't exist at all
+    pub fn open(path: &str, create_if_missing: bool) -> Result<Self, Error>{
+        let db_path = path.to_owned() + Self::DB_DIR;
+
+        // Preparing the CurrentState DB directory if it doesn't exist
+        let db_path_exists = Path::new(db_path.as_str()).exists();
+        if !db_path_exists {
+            if !create_if_missing {
+                return Err(Error::new("No need to create a DB".into()));
+            } else {
+                if std::fs::create_dir_all(&db_path).is_err(){
+                    return Err(Error::new("DB directory can't be created".into()))
+                }
+            }
+        }
+
+        let mut opts = Options::default();
+        opts.create_if_missing(create_if_missing);
+
+        // Opening or creating the CurrentState DB
+        let db =
+            if db_path_exists {
+                TransactionDB::open_cf_all(&opts, &db_path)?
+            } else {
+                TransactionDB::open_cf_default(&opts, &db_path)?
+            };
+
+        // Checking or creating the Versions directory
+        let versions_path = path.to_owned() + Self::VERSIONS_DIR;
+        if Path::new(versions_path.as_str()).exists(){
+            // Scan all existing versions and check that their numbers are a consecutive set
+            Self::get_all_versions_in_dir(versions_path.as_str())?;
+        } else {
+            // Creating the Versions directory if it doesn't exist
+            if std::fs::create_dir(&versions_path).is_err(){
+                drop(db);
+                return if std::fs::remove_dir(&db_path).is_ok() {
+                    Err(Error::new("Versions directory can't be created".into()))
+                } else {
+                    Err(Error::new("Versions directory can't be created; Can't clean the DB directory".into()))
+                }
+            }
+        }
+
+        Ok(
+            Self{
+                db,
+                db_path: Self::absolute_path(&db_path)?,
+                versions_path: Self::absolute_path(&versions_path)?,
+                base_path: Self::absolute_path(&path.to_owned())?
+            }
+        )
+    }
+
+    // Creates a transaction for a current state of storage if 'version_id_opt' is 'None', or for a specified previous version of the storage otherwise.
+    // Returns Result with TransactionVersioned or with Error message if some error occurred
+    pub fn create_transaction(&self, version_id_opt: Option<&str>) -> Result<TransactionVersioned, Error> {
+        if let Some(version_id) = version_id_opt {
+            let db_version = self.open_version(version_id)?;
```

Comment:
what if version will be the same as the last version? Is it fine to open the same DB twice?

Reply:
If the version is the same as the last saved version then it will be opened as well as any other existing version; 
If the DB with version is already opened, then an attempt to open it twice will fail unless the transaction for this version (which contains a handle to the DB) is closed (dropped).